codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -16,329 +16,514 @@
+        encoder().writeHeaders(ctx, streamId, headersFrame.headers(), headersFrame.padding(),
+                                            headersFrame.isEndStream(), writePromise);
     }
 
+    // TODO: Should we maybe also send Http2FrameStreamEvents for all the other stream state changes ?
     private final class ConnectionListener extends Http2ConnectionAdapter {
+
         @Override
         public void onStreamActive(Http2Stream stream) {
-            if (ctx == null) {
-                // UPGRADE stream is active before handlerAdded().
-                return;
-            }
-            if (isOutboundStream(server, stream.id())) {
-                // Creation of outbound streams is notified in writeHeadersFrame().
+            if (isOutboundStream(connection().isServer(), stream.id())) {
                 return;
             }
-            ctx.fireUserEventTriggered(new Http2StreamActiveEvent(stream.id()));
+
+            DefaultHttp2FrameStream stream2 = newStream().attach(streamKey, stream);
+            onHttp2StreamStateChanged(ctx, stream2);
         }
 
         @Override
         public void onStreamClosed(Http2Stream stream) {
-            ctx.fireUserEventTriggered(new Http2StreamClosedEvent(stream.id()));
+            DefaultHttp2FrameStream stream2 = stream.getProperty(streamKey);
+            if (stream2 != null) {
+                onHttp2StreamStateChanged(ctx, stream2);
+            }
+        }
+
+        @Override
+        public void onStreamHalfClosed(Http2Stream stream) {
+            DefaultHttp2FrameStream stream2 = stream.getProperty(streamKey);
+            if (stream2 != null) {
+                onHttp2StreamStateChanged(ctx, stream2);
+            }
         }
 
         @Override
         public void onGoAwayReceived(final int lastStreamId, long errorCode, ByteBuf debugData) {
-            ctx.fireChannelRead(new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData.retain()));
+            onHttp2Frame(ctx, new DefaultHttp2GoAwayFrame(lastStreamId, errorCode, debugData.retain()));
         }
     }
 
-    private static final class InternalHttp2ConnectionHandler extends Http2ConnectionHandler {
-        InternalHttp2ConnectionHandler(Http2ConnectionDecoder decoder, Http2ConnectionEncoder encoder,
-                                       Http2Settings initialSettings) {
-            super(decoder, encoder, initialSettings);
+    @Override
+    protected final void onConnectionError(ChannelHandlerContext ctx, Throwable cause, Http2Exception http2Ex) {
+        ctx.fireExceptionCaught(cause);
+    }
+
+    /**
+     * Exceptions for unknown streams, that is streams that have no {@link Http2FrameStream} object attached
+     * are simply logged and replied to by sending a RST_STREAM frame. There is not much value in propagating such
+     * exceptions through the pipeline, as a user will not have any additional information / state about this
+     * stream and thus can't do any meaningful error handling.
```

Comment:
I added a `onHttp2UnknownStreamError(...)` method that we can override if we really need to later on.